src/validation.h

```diff
@@ -711,14 +732,170 @@ class SnapshotMetadata
 
 };
 
+/**
+ * Provides an interface for creating and interacting with multiple
+ * chainstates, as well as a means to initialize chainstates from a
+ * UTXO snapshot. Managed chainstates can be maintained at different
+ * heights simultaneously.
+ *
+ * This class provides abstractions that allow the retrieval of the current
+ * most-work chainstate ("Active") as well as chainstates which may be in
+ * background use to validate UTXO snapshots.
+ *
+ * This class was created to displace the use of global variables (e.g.
+ * chainActive) which were historically used to reference a single active
+ * chainstate throughout the codebase.
+ *
+ * TODO jamesob: basically everything in this class is implicitly covered
+ *   by `cs_main`. Should probably put some more formalization around
+ *   locking and maybe add a class-local `cs` CCriticalSection.
+ */
+class ChainstateManager
+{
+private:
+    /**
+     * Temporary holding place for a snapshot chainstate which is in the process of
+     * being constructed. When ready, we swap with `m_snapshot_chainstate`.
+     */
+    std::unique_ptr<CChainState> m_unready_snapshot_chainstate;
+
+    /**
+     * If not empty, this points to metadata corresponding to the UTXO snapshot
+     * currently in use.
+     */
+    std::unique_ptr<SnapshotMetadata> m_snapshot_metadata;
+
+    /**
+     * The chainstate used under normal operation (i.e. "regular" IBD) or,
+     * if a snapshot is in use, for background validation. Its contents will
+     * be freed when background validation of the snapshot has completed.
+     */
+    std::unique_ptr<CChainState> m_ibd_chainstate GUARDED_BY(cs_main);
+
+    /**
+     * A chainstate initialized on the basis of a UTXO snapshot. If this is
+     * truthy, it is always our active chainstate unless proven invalid.
```

Comment:
In commit "validation: introduce ChainstateManager, use in init" (8d9b707f609b647fb28ae182a097adc03e718725)

Probably clearer to say "set" or "non-null" here instead of "truthy".